src/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/StartupCode/StartupCodeHelpers.cs

```diff
@@ -63,43 +57,24 @@ internal static unsafe void InitializeCommandLineArgs(int argc, byte** argv)
             Environment.SetCommandLineArgs(args);
         }
 
-        private static unsafe IntPtr[] CreateModuleManagers()
+        private static unsafe IntPtr[] CreateModuleManagers(IntPtr moduleHeaders, int count)
         {
-            IntPtr* modulesSectionStart;
-            IntPtr* modulesSectionEnd;
-            GetModulesSection(out modulesSectionStart, out modulesSectionEnd);
-
             // Count the number of modules so we can allocate an array to hold the ModuleManager objects.
             // At this stage of startup, complex collection classes will not work.
             int moduleCount = 0;
-            for (IntPtr* current = modulesSectionStart; current < modulesSectionEnd; current++)
+            for (int i = 0; i < count; i++)
             {
-                IntPtr* moduleHeaderStart = (IntPtr*)*current;
-
-                if (moduleHeaderStart != null)
-                {
+                // TODO: Why are we getting the zero pointers in the list?
```

Comment:
I think this is a result of the PE section merging trick and merging the sections .modules$A, .modules$Z with each module's .modules$I section in between. We end up with ~0x100 of padding / alignment for each of the three merged sections. I stepped through the C++ runtime startup wondering the same thing a while back and they see the same padding with their cctor initialization code. I don't think we can do better with #pragmas but it might be possible to use the /align switch to the linker to reduce the padding.